### DISCUSSION TITLE
Custom ExceptionMiddleware not catching exceptions

### DISCUSSION THREAD

**Discussion by mshann03** [ChilliCream/hotchocolate#5510] [Q&A] [answered]
Hi @michaelstaib,

I'm new to HotChocolate and GraphQL, but I'm loving what your library can do.

I have a basic requirement where I want to use FluentValidation, and if a ValidationException is thrown, then I use the error builder to transform those error details and throw a GraphQLException. Something like:

```
public class ValidationMiddleware
{
    private readonly HotChocolate.Execution.RequestDelegate _next;

    public ValidationMiddleware(HotChocolate.Execution.RequestDelegate next)
    {
        _next = next;
    }

   public async ValueTask InvokeAsync(IRequestContext context)
    {
        try
        {
            await _next(context);
        }
        catch (ValidationException ve)
        {
            var errors = ve.Errors.Select(e =>
                ErrorBuilder
                    .New()
                    .SetCode(e.ErrorCode)
                    .SetMessage(e.ErrorMessage)
                    .SetExtension("Property", e.PropertyName)
                    .Build());

            throw new GraphQLException(errors);
        }
    }
}
```

I was hoping it would then get caught by your ExceptionMiddleware and output accordingly. My program setup looks like:

```
builder.Services
    .AddGraphQLServer()
    .AddQueryType<Query>()
    .UseInstrumentations()
    .UseRequest<ValidationMiddleware>()
    .UseExceptions()
    .UseTimeout()
    .UseDocumentCache()
    .UseDocumentParser()
    .UseDocumentValidation()
    .UseOperationCache()
    .UseOperationComplexityAnalyzer()
    .UseOperationResolver()
    .UseOperationVariableCoercion()
    .UseOperationExecution();
```

However, my middleware always completes fine without any exceptions caught. Any idea what I'm doing wrong?

Here is a silly example query to illustrate what I'm doing:

```
public class Something
 {
      public int MyProperty { get; set; }
 }

public class Query
{
    public async Task<Something> GetSomething()
    {
        throw new ValidationException("Something went wrong!");
    }
}
```

**Comment by mshann03** [comment]
Hi All, does anyone have a solution for this? It would be nice to know what I'm doing wrong here. Thanks!

**Comment by venkat-bhogavilli** [comment]
HI .... what I think is your middleware is being registered very late in request execution pipeline. You can try to use immediately after AddGraphQLServer like below ......or 
even try to intercept HTTPRequest like https://chillicream.com/docs/hotchocolate/v12/server/interceptors 
Hope this helps

```
builder.Services
    .AddGraphQLServer()
    .UseRequest<ValidationMiddleware>()
    .AddQueryType<Query>()
    .UseInstrumentations()
    .UseExceptions()
    .UseTimeout()
    .UseDocumentCache()
    .UseDocumentParser()
    .UseDocumentValidation()
    .UseOperationCache()
    .UseOperationComplexityAnalyzer()
    .UseOperationResolver()
    .UseOperationVariableCoercion()
    .UseOperationExecution();
```

**Comment by michaelstaib** [comment, marked as answer]
Resolver error are not propagated to the request middleware... 

The request pipeline represents a modular pipeline for the request level execution steps.

The field middleware represents this for the GraphQL field. In your case you are throwing an exception in the resolver so you need to use a filed middleware. 

In any case, if you just want to catch and log resolver level errors / exception use the diagnostic events an override ResolverError.
